vms/evm/uptimetracker/state.go

```diff
@@ -0,0 +1,234 @@
+const codecVersion uint16 = 0
+
+var (
+	codecManager codec.Manager
+	_            uptime.State = (*state)(nil)
+)
+
+func init() {
+	codecManager = codec.NewManager(math.MaxInt32)
+	c := linearcodec.NewDefault()
+
+	if err := c.RegisterType(validator{}); err != nil {
+		panic(fmt.Errorf("failed to register type: %w", err))
+	}
+
+	if err := codecManager.RegisterCodec(codecVersion, c); err != nil {
+		panic(fmt.Errorf("failed to register codec: %w", err))
+	}
+}
+
+type validator struct {
+	UpDuration    time.Duration `serialize:"true"`
+	LastUpdated   uint64        `serialize:"true"`
+	NodeID        ids.NodeID    `serialize:"true"`
+	Weight        uint64        `serialize:"true"`
+	StartTime     uint64        `serialize:"true"`
+	IsActive      bool          `serialize:"true"`
+	IsL1Validator bool          `serialize:"true"`
+
+	validationID ids.ID
+}
+
+// state holds the on-disk and cached representation of the validator set
+type state struct {
+	db database.Database
+
+	validators             map[ids.ID]*validator
+	nodeIDsToValidationIDs map[ids.NodeID]ids.ID
+	updatedValidators      set.Set[ids.ID]
+	deletedValidators      set.Set[ids.ID]
+}
+
+func newState(db database.Database) (*state, error) {
+	s := &state{
+		db:                     db,
+		validators:             make(map[ids.ID]*validator),
+		nodeIDsToValidationIDs: make(map[ids.NodeID]ids.ID),
+	}
+
+	it := db.NewIterator()
+	defer it.Release()
+
+	for it.Next() {
+		validationID, err := ids.ToID(it.Key())
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse validation ID: %w", err)
+		}
+
+		vdr := &validator{
+			validationID: validationID,
+		}
+
+		if _, err := codecManager.Unmarshal(it.Value(), vdr); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal validator: %w", err)
+		}
+
+		s.addValidator(validationID, vdr)
+	}
+
+	if err := it.Error(); err != nil {
+		return nil, fmt.Errorf("failed to iterate: %w", err)
+	}
+
+	return s, nil
+}
+
+func (s *state) GetUptime(nodeID ids.NodeID) (time.Duration, time.Time, error) {
+	v, ok := s.getValidatorByNodeID(nodeID)
+	if !ok {
+		return 0, time.Time{}, database.ErrNotFound
+	}
+
+	return v.UpDuration, time.Unix(int64(v.LastUpdated), 0), nil
+}
+
+func (s *state) SetUptime(
+	nodeID ids.NodeID,
+	upDuration time.Duration,
+	lastUpdated time.Time,
+) error {
+	v, ok := s.getValidatorByNodeID(nodeID)
+	if !ok {
+		return database.ErrNotFound
+	}
+
+	v.UpDuration = upDuration
+	v.LastUpdated = uint64(lastUpdated.Unix())
+
+	s.updatedValidators.Add(v.validationID)
+
+	return nil
+}
+
+func (s *state) GetStartTime(nodeID ids.NodeID) (time.Time, error) {
+	v, ok := s.getValidatorByNodeID(nodeID)
+	if !ok {
+		return time.Time{}, database.ErrNotFound
+	}
+
+	return time.Unix(int64(v.LastUpdated), 0), nil
+}
+
+func (s *state) addNewValidator(vdr *validator) {
+	s.addValidator(vdr.validationID, vdr)
+	s.updatedValidators.Add(vdr.validationID)
+}
+
+func (s *state) updateValidator(validationID ids.ID, isActive bool) bool {
```

Comment:
Should we document that this function assumes the validator currently exists?

If this is called when the validator doesn't exist it will panic... Which is different than `deleteValidator`